ouroboros-consensus-test/src/Test/Util/Serialisation/Golden.hs

```diff
@@ -464,48 +468,80 @@ goldenTest_SerialiseNodeToClient codecConfig goldenDir Examples {..} =
   FlatTerm
 -------------------------------------------------------------------------------}
 
-deriving instance Generic TermToken
-deriving instance ToExpr  TermToken
-
 instance Condense TermToken where
   condense = show
 
 {-------------------------------------------------------------------------------
   Diffing Cbor
 -------------------------------------------------------------------------------}
 
-data CborTree =
-    FlatTerm   TermToken
-  | NestedTerm CborForest
-
-type CborForest = [CborTree]
+diffToExpr :: ToExpr a => a -> a -> String
+diffToExpr x y = show (ansiWlEditExpr (ediff x y))
+
+newtype CBORBytes = CBORBytes Strict.ByteString
+
+instance ToExpr CBORBytes where
+  toExpr (CBORBytes bytes) =
+    case CBOR.deserialiseFromBytes CBOR.decodeTerm $ Lazy.fromStrict bytes of
+      Left err -> error $ "Error decoding CBOR: " ++ show err
+      Right (bytesLeftOver, term)
+        | Lazy.null bytesLeftOver -> termToExpr term
+        | otherwise ->
+          error $
+          unlines
+            [ "Unexpected leftover bytes: "
+            , show $ showHexBytesGrouped $ Lazy.toStrict bytesLeftOver
+            , "when decoding"
+            , show term
+            ]
+    where
+      hexByteString bs =
+        [ toExpr (Strict.length bs)
+        , Lst (map toExpr $ showHexBytesGrouped bs)
+        ]
+      termToExpr =
+        \case
+          CBOR.TInt i -> App "TInt" [toExpr i]
+          CBOR.TInteger i -> App "TInteger" [toExpr i]
+          CBOR.TBytes bs -> App "TBytes" $ hexByteString bs
+          CBOR.TBytesI bs -> App "TBytesI" $ hexByteString $ Lazy.toStrict bs
+          CBOR.TString s -> App "TString" [toExpr s]
+          CBOR.TStringI s -> App "TStringI" [toExpr s]
+          CBOR.TList xs -> App "TList" [Lst (map termToExpr xs)]
+          CBOR.TListI xs -> App "TListI" [Lst (map termToExpr xs)]
+          CBOR.TMap xs ->
+            App "TMap" [Lst (map (toExpr . bimap termToExpr termToExpr) xs)]
+          CBOR.TMapI xs ->
+            App "TMapI" [Lst (map (toExpr . bimap termToExpr termToExpr) xs)]
+          CBOR.TTagged 24 (CBOR.TBytes x) ->
+            App "CBOR-in-CBOR" [toExpr (CBORBytes x)]
+          CBOR.TTagged t x -> App "TTagged" [toExpr t, termToExpr x]
+          CBOR.TBool x -> App "TBool" [toExpr x]
+          CBOR.TNull -> App "TNull" []
+          CBOR.TSimple x -> App "TSimple" [toExpr x]
+          CBOR.THalf x -> App "THalf" [toExpr x]
+          CBOR.TFloat x -> App "TFloat" [toExpr x]
+          CBOR.TDouble x -> App "TDouble" [toExpr x]
+
+
+
+-- | Show a ByteString as hex groups of 8bytes each. This is a slightly more
+-- useful form for debugging, rather than bunch of escaped characters.
+showHexBytesGrouped :: Strict.ByteString -> [String]
+showHexBytesGrouped bs = reverse (go [] (showsHexBytes bs))
+  where
+    go acc [] = acc
+    go acc (x1:x2:x3:x4:x5:x6:x7:x8:xs) =
+      acc `seq`
```

Comment:
I've changed implementation to not use any recursion, so these comments are no longer relevant, but useful nevertheless. Thank you.

> we generally prefer -XBangPatterns to explicit seq

Sounds good. I'll keep it in mind for the future.

>  this seq seems unnecessary: the accumulator is always in WHNF by construction in this loop

Yes, the WHNF was an oversight, it has to be NH, so `deepseq` would have taken care of the memory leak:

```haskell
    go !acc =
      \case
        [] -> acc
        (x1:x2:x3:x4:x5:x6:x7:x8:xs) ->
          let hex = ("0x" ++) . x1 . x2 . x3 . x4 . x5 . x6 . x7 . x8 $ ""
          in hex `deepseq` go (hex : acc) xs
        xs -> foldr ($) "" (("0x" ++) : xs) : acc
```

> And the last go case isn't strict in acc

I am not sure that matters much, since it is just one of the two base cases and is not part of the inner loop. However, as I mentioned earlier, It isn't really relevant now anyways :)